std/std.sol

```diff
@@ -34,184 +34,193 @@ function snd(p: Pair(a, b)) -> b {
-
+        return Typedef.abs(res);
     }
+}
 
-    function store(loc: Memory(t), value: t) -> Unit {
-        let rw = rep(loc);
-        let vw = rep(value);
+instance memory(t) : Storable(t) {
+    function store(loc: memory(t), value: t) -> Unit {
+        let rw : word = Typedef.rep(loc);
+        let vw : word = Typedef.rep(value);
         assembly {
           mstore(rw, vw)
         };
     }
 }
 
+instance memory(t) : Ref(t) {}
 
+// Free Memory Pointer
 
-// Memory Layout
-
-function get_free_memory() -> Word {
-    let res : Word;
+function get_free_memory() -> word {
+    let res : word;
     assembly {
         res := mload(0x40)
     };
     return res;
 }
 
-function set_free_memory(loc : Word) {
+function set_free_memory(loc : word) {
     assembly {
         mstore(0x40, loc)
     };
 }
 
 
-// Bytes
+// bytes
 
-data Bytes = Bytes
 // TODO: bytes should be indexable (but not ref since it can't live on stack...)
+// TODO: we would really want this to be a typedef to void (or a datatype with no constructors)
+data bytes = bytes;
 
 // ABI Encoding
 
 class t:EncodeInto {
-    function encodeInto(val : t, hd : Word, tl : Word) -> Pair(Word,Word);
+    function encodeInto(val : t, hd : word, tl : word) -> Pair(word,word);
 }
 
 class t:Encode {
     // TODO: is the following ever needed??
     // is the abi representation of `t` dynamically sized?
-    // function isDynamicallySized(x:Proxy(t)) -> Bool;
+    // function isDynamicallySized(x:Proxy(t)) -> bool;
 
     // does `t` (or any of it's children) contain a dynamic type?
-    function shouldEncodeDynamic(x:Proxy(t)) -> Bool;
+    function shouldEncodeDynamic(x:Proxy(t)) -> bool;
     // how big is the head portion of the ABI representation of `t`?
-    function headSize(x:Proxy(t)) -> Word;
+    function headSize(x:Proxy(t)) -> word;
 }
 
-function encode(val:t) -> Memory(Bytes) {
-    let hdSz = headSize(Proxy);
-    let ptr = get_free_memory();
-    let head: Word;
-    let tail: Word;
+forall t:Encode, t:EncodeInto . function encode(val:t) -> memory(bytes) {
+    let p : Proxy(t);
+    let hdSz = Encode.headSize(p);
+    let ptr : word = get_free_memory();
+    let head : word;
+    let tail : word;
     assembly {
         head := add(ptr, 32);
         tail := add(head, hdSz);
     };
-    let tl = snd(encodeInto(val,head,tail));
+    let tl = snd(EncodeInto.encodeInto(val,head,tail));
     set_free_memory(tl);
     assembly {
         mstore(ptr, sub(tl, head))
     };
-    return abs(ptr);
+    let mem : memory(bytes) = memory(ptr);
+    return mem;
 }
 
 instance (l:Encode, r:Encode) => Pair(l,r):Encode {
-    function shouldEncodeDynamic(x : Proxy(Pair(l,r))) -> Bool {
+    function shouldEncodeDynamic(x : Proxy(Pair(l,r))) -> bool {
         let l: Proxy(l) = Proxy;
         let r: Proxy(r) = Proxy;
-        return and(shouldEncodeDynamic(l), shouldEncodeDynamic(l));
+        return and(Encode.shouldEncodeDynamic(l), Encode.shouldEncodeDynamic(l));
```

Comment:
I just scrolled through the code.. isn't this a typo here? shouldn't this expression be `and(Encode.shouldEncodeDynamic(l), Encode.shouldEncodeDynamic(r));`?